packages/block-editor/src/utils/style-variable-resolution.js

```diff
@@ -0,0 +1,180 @@
+/**
+ * External dependencies
+ */
+import { get, find, isString, kebabCase } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { __EXPERIMENTAL_PRESET_METADATA as PRESET_METADATA } from '@wordpress/blocks';
+
+const STYLE_PROPERTIES_TO_CSS_VAR_INFIX = {
+	linkColor: 'color',
+	backgroundColor: 'color',
+	background: 'gradient',
+};
+
+function findInPresetsBy(
+	features,
+	context,
+	presetPath,
+	presetProperty,
```

Comment:
How about naming these two `presetPropertyName`, `presetPropertyValue` to communicate what they are?

Reply:
It seems the role of `findInPresetsBy` is to find the preset object whose `presetPropertyName` matches the `presetPropertyValue` among all the available.

The complexity here is that it needs to take into account that 1) there may be a block and a global preset and 2) any of those may have core, theme, and user presets. This is the full list of candidate presets available, sorted by priority:

1. block preset from user
2. block preset from theme
3. block preset from core
4. global preset from user
5. global preset from theme
6. global preset from core

My understanding is that we should return the first candidate whose `presetPropertyName` matches `presetPropertyValue`. Wouldn't that be enough? I've looked at the PR that introduced this function https://github.com/WordPress/gutenberg/pull/33149 and I don't understand why we need the recursion when the property to look up is not the `slug`.

Reply:
I've updated in `trunk` the implementation of this function to this algorithm:

1. Create a list of the available presets, sorted by priority.
2. Return the first preset.

And it works as described at https://github.com/WordPress/gutenberg/pull/33149

Code, for reference:

```js
function findInPresetsBy(
	styles,
	blockName,
	presetPath,
	presetPropertyName,
	presetPropertyValue
) {
	const candidates = [];
	const origins = [ 'user', 'theme', 'user' ];

	const blockPresets = get( styles, [
		'settings',
		'blocks',
		blockName,
		...presetPath,
	] );
	origins.forEach( ( origin ) => {
		if ( blockPresets?.[ origin ] ) {
			blockPresets[ origin ].forEach( ( preset ) => {
				if ( preset[ presetPropertyName ] === presetPropertyValue ) {
					candidates.push( preset );
				}
			} );
		}
	} );

	const globalPresets = get( styles, [ 'settings', ...presetPath ] );
	origins.forEach( ( origin ) => {
		if ( globalPresets?.[ origin ] ) {
			globalPresets[ origin ].forEach( ( preset ) => {
				if ( preset[ presetPropertyName ] === presetPropertyValue ) {
					candidates.push( preset );
				}
			} );
		}
	} );

	return candidates.length >= 1 ? candidates[ 0 ] : undefined;
}
```

Reply:
Hi @oandregal,
> My understanding is that we should return the first candidate whose presetPropertyName matches presetPropertyValue. Wouldn't that be enough?

I don't think it is enough.
Imagining user theme.json defines presets:
```
{
	color: "#000"
	slug: "black",
	name: "Black"
}
```
And theme defines: 
{
	color: "#001"
	slug: "black",
	name: "Black"
}

If we search the color property value of "#001" without the recursion we would return the theme preset, but we should not return it because the variable was overwritten and now there is no variable with color "#001".
The recursion makes sure that when we find a preset, there is no other higher priority preset with the same slug and different property because if there is the preset should be ignored.

Reply:
Nice one :+1: I've pushed some tests in https://github.com/WordPress/gutenberg/pull/34178/commits/c78a32292b54d357d4c570ac34c2dc1605e6017c to make sure this behavior is documented and retained.